HLTrigger/Timer/src/FastTimerService.cc

```diff
@@ -38,1425 +39,1365 @@
-    for (auto const & name: tns.getTrigPaths()) {
-      out << "FastReport              "
-          << std::right << std::setw(10) << m_stream.paths[pid][name].summary_active        / (double) summary.count << " "
-          << std::right << std::setw(10) << m_stream.paths[pid][name].summary_premodules    / (double) summary.count << " "
-          << std::right << std::setw(10) << m_stream.paths[pid][name].summary_intermodules  / (double) summary.count << " "
-          << std::right << std::setw(10) << m_stream.paths[pid][name].summary_postmodules   / (double) summary.count << " "
-          << std::right << std::setw(10) << m_stream.paths[pid][name].summary_overhead      / (double) summary.count << " "
-          << std::right << std::setw(10) << m_stream.paths[pid][name].summary_total         / (double) summary.count << "  "
-          << name << '\n';
-    }
-    out << '\n';
-    //out << "FastReport " << (m_use_realtime ? "(real time) " : "(CPU time)  ")    << "     Active       Pre-     Inter-  Post-mods   Overhead      Total  EndPath" << '\n';
-    for (auto const & name: tns.getEndPaths()) {
-      out << "FastReport              "
-          << std::right << std::setw(10) << m_stream.paths[pid][name].summary_active        / (double) summary.count << " "
-          << std::right << std::setw(10) << m_stream.paths[pid][name].summary_premodules    / (double) summary.count << " "
-          << std::right << std::setw(10) << m_stream.paths[pid][name].summary_intermodules  / (double) summary.count << " "
-          << std::right << std::setw(10) << m_stream.paths[pid][name].summary_postmodules   / (double) summary.count << " "
-          << std::right << std::setw(10) << m_stream.paths[pid][name].summary_overhead      / (double) summary.count << " "
-          << std::right << std::setw(10) << m_stream.paths[pid][name].summary_total         / (double) summary.count << "  "
-          << name << '\n';
-    }
-  }
-  out << '\n';
-  if (m_enable_timing_modules) {
-    //out << "FastReport " << (m_use_realtime ? "(real time) " : "(CPU time)  ")    << "     Active  Module" << '\n';
-    for (auto & keyval: m_stream.modules) {
-      std::string const & label  = keyval.first;
-      ModuleInfo  const & module = keyval.second;
-      out << "FastReport              " << std::right << std::setw(10) << module.summary_active  / (double) summary.count << "  " << label << '\n';
-    }
-    //out << "FastReport " << (m_use_realtime ? "(real time) " : "(CPU time)  ")    << "     Active  Module" << '\n';
-    out << '\n';
-    //out << "FastReport " << (m_use_realtime ? "(real time) " : "(CPU time)  ")    << "     Active  Module" << '\n';
-    for (auto & keyval: m_stream.moduletypes) {
-      std::string const & label  = keyval.first;
-      ModuleInfo  const & module = keyval.second;
-      out << "FastReport              " << std::right << std::setw(10) << module.summary_active  / (dou
```

Comment:
@Dr15Jones turns out that this is not good enough :-(

In a job configured to run with N threads, I have sometimes seen an (N+1)th or (N+2)th thread appear;  `threadId()` nicely assigns them new ids, but then `threads_` is not big enough.

Here in an excerpt with some debugging messages added in:
```
%MSG-i ThreadSetup:  (NoModuleName) 22-Apr-2017 01:11:27 CEST pre-events
setting # threads 8
%MSG
%MSG-i StreamSetup:  (NoModuleName) 22-Apr-2017 01:11:27 CEST pre-events
setting # streams 8
%MSG
...
thread 139795564551296 assigned unique id 0
thread 139795564551296 using unique id 0 / 7
thread 139795564551296 using unique id 0 / 7
...
thread 139793846748928 assigned unique id 1
thread 139793846748928 using unique id 1 / 7
thread 139793899976448 assigned unique id 2
thread 139793899976448 using unique id 2 / 7
thread 139793899976448 using unique id 2 / 7
thread 139793846748928 using unique id 1 / 7
thread 139793846748928 using unique id 1 / 7
thread 139793867728640 assigned unique id 4
thread 139793650677504 assigned unique id 5
thread 139793889486592 assigned unique id 3
thread 139793640187648 assigned unique id 7
thread 139793857238784 assigned unique id 6
thread 139793650677504 using unique id 5 / 7
thread 139793867728640 using unique id 4 / 7
...
```
and then much later
```
...
thread 139793878218496 assigned unique id 8
thread 139793878218496 using unique id 8 / 7
thread 139793835489024 assigned unique id 9
thread 139793889486592 using unique id 3 / 7
thread 139793889486592 using unique id 3 / 7
thread 139793889486592 using unique id 3 / 7
thread 139793889486592 using unique id 3 / 7
thread 139793889486592 using unique id 3 / 7
thread 139793835489024 using unique id 9 / 7
...
```
soon after leading to
```
----- Begin Fatal Exception 22-Apr-2017 01:13:10 CEST-----------------------
An exception of category 'StdException' occurred while
   [0] Processing  Event run: 165121 lumi: 62 event: 23684942 stream: 4
   [1] Running path 'pathALCARECODtCalib'
   [2] Prefetching for module GoodVertexFilter/'primaryVertexFilter'
   [3] Prefetching for module RecoChargedRefCandidatePrimaryVertexSorter/'offlinePrimaryVertices'
   [4] Prefetching for module FastjetJetProducer/'ak4CaloJetsForTrk'
   [5] Prefetching for module CaloTowersCreator/'caloTowerForTrk'
   [6] Prefetching for module EcalRecHitProducer/'ecalRecHit'
   [7] Prefetching for module EcalUncalibRecHitProducer/'ecalMultiFitUncalibRecHit'
   [8] Calling method for module EcalRawToDigi/'ecalDigis'
Exception Message:
A std::exception was thrown.
vector::_M_range_check: __n (which is 9) >= this->size() (which is 8)
----- End Fatal Exception -------------------------------------------------
```

Do you have any ideas what might be going on ?
So far I have thought of
  - TBB is removing old threads and adding new threads to the pool (possibility suggested by the existence of a `tbb:task_scheduler_observer` class, see https://software.intel.com/en-us/node/506314);
  - some of the signals end up on threads that are handling xrootd or other I/O (I don't think I have see this happen with an `EmptysSurce`).

Either way, the workarounds I could think of are
  - use a `thread local Measurement` instead of a `std::vector<Measurement>`: this will give static objects which would not be suited for a normal module, but since a Service is kind of a singleton anyway, it could work;
  - use a `tbb::concurrent_hash_map<thread::id,Measurement>` keyed on `this_thread::get_id()`: this should work, no idea on its efficiency;
  - use a `tbb::enumerable_thread_specific<Measurement>`: this should do the same as the previous one, but is specialised for this use case.
 
Comments or suggestions ?